include/xtensor/xgenerator.hpp

```diff
@@ -0,0 +1,494 @@
+        const_reference operator[](const xindex& index) const;
+
+        template <class It>
+        const_reference element(It first, It last) const;
+
+        template <class S>
+        bool broadcast_shape(S& shape) const;
+
+        template <class S>
+        bool is_trivial_broadcast(const S& /*strides*/) const noexcept;
+
+        const_iterator begin() const noexcept;
+        const_iterator end() const noexcept;
+        const_iterator cbegin() const noexcept;
+        const_iterator cend() const noexcept;
+
+        template <class S>
+        xiterator<const_stepper, S> xbegin(const S& shape) const noexcept;
+        template <class S>
+        xiterator<const_stepper, S> xend(const S& shape) const noexcept;
+        template <class S>
+        xiterator<const_stepper, S> cxbegin(const S& shape) const noexcept;
+        template <class S>
+        xiterator<const_stepper, S> cxend(const S& shape) const noexcept;
+
+        template <class S>
+        const_stepper stepper_begin(const S& shape) const noexcept;
+
+        template <class S>
+        const_stepper stepper_end(const S& shape) const noexcept;
+
+        const_storage_iterator storage_begin() const noexcept;
+        const_storage_iterator storage_end() const noexcept;
+
+        const_storage_iterator storage_cbegin() const noexcept;
+        const_storage_iterator storage_cend() const noexcept;
+
+    private:
+
+        template <class Func, std::size_t... I>
+        const_stepper build_stepper(Func&& f, std::index_sequence<I...>) const noexcept;
+
+        functor_type m_f;
+        shape_type m_shape;
+        friend class xgenerator_stepper<F, R>;
+    };
+
+    /***************************
+     * xgenerator_stepper *
+     ***************************/
+
+    template <class F, class R>
+    class xgenerator_stepper
+    {
+
+    public:
+
+        using self_type = xgenerator_stepper<F, R>;
+        using functor_type = typename std::remove_reference<F>::type;
+        using xfunction_type = xgenerator<F, R>;
+
+        using value_type = typename xfunction_type::value_type;
+        using reference = typename xfunction_type::value_type;
+        using pointer = typename xfunction_type::const_pointer;
+        using size_type = typename xfunction_type::size_type;
+        using difference_type = typename xfunction_type::difference_type;
+        using iterator_category = std::input_iterator_tag;
+
+        using shape_type = typename xfunction_type::shape_type;
+
+        xgenerator_stepper(const xfunction_type* func, const shape_type& shape) noexcept;
+
+        void step(size_type dim, size_type n = 1);
+        void step_back(size_type dim, size_type n = 1);
+        void reset(size_type dim);
+
+        void to_end();
+
+        reference operator*() const;
+
+        bool equal(const self_type& rhs) const;
+
+    private:
+        const xfunction_type* p_f;
+        shape_type m_shape;
+        shape_type m_index;
```

Comment:
You can use the `get_index_type` meta function to deduce the `index_type` from `shape_type`.